File: dailyQuestion/2020-07/07-28/python/solution_bfs.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def maxDepth(root: TreeNode) -> int:
    if not root:
        return 0
    height = 0
    queue = [root]
    while queue:
        size = len(queue)
        height += 1
        stack = []
        while size != 0:
            node = queue.pop()
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
            size -= 1
        queue.extend(stack)
    return height



if __name__ == "__main__" :
    node = TreeNode(3)
    node.left = TreeNode(9)
    node.left.left = TreeNode(5)
    node.left.right = TreeNode(8)
    node.right = TreeNode(20)
    node.right.left = TreeNode(15)
    node.right.right = TreeNode(7)
    result = maxDepth(node)
    print(result)

File: dailyQuestion/2020/2020-07/07-28/golang/solution_bfs.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	queue := []*TreeNode{root}
	height := 0
	for len(queue) != 0 {
		height++
		size := len(queue)
		stack := []*TreeNode{}
		for size != 0 {
			node := queue[size-1]
			if node.Left != nil {
				stack = append(stack, node.Left)
			}
			if node.Right != nil {
				stack = append(stack, node.Right)
			}
			queue = queue[:size-1]
			size--
		}
		queue = append(queue, stack...)
	}
	return height
}

func main() {
	root := &TreeNode{Val: 3, Left: &TreeNode{Val: 5}, Right: &TreeNode{Val: 1}}
	root.Left.Left = &TreeNode{Val: 6}
	root.Left.Right = &TreeNode{Val: 2, Left: &TreeNode{Val: 7}, Right: &TreeNode{Val: 4}}
	root.Right.Left = &TreeNode{Val: 0}
	root.Right.Right = &TreeNode{Val: 8}
	result := maxDepth(root)
	fmt.Println(result)
}
